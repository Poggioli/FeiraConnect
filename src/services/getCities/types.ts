import { PaginationRequest, PaginationResponse } from "@/services/api";
import { InfiniteData, UseInfiniteQueryOptions } from "@tanstack/react-query";
import { AxiosError } from "axios";

export type GetCities = {
  id: string,
  slug: string;
  name: string;
};

export type GetCitiesParams = PaginationRequest<{
  q: string;
}>;

export type GetCitiesResponse = PaginationResponse<GetCities>;

export type UseGetCitiesOptions = Omit<
  UseInfiniteQueryOptions<
    GetCitiesResponse,
    AxiosError,
    InfiniteData<GetCitiesResponse>,
    GetCitiesResponse,
    ["getCities", string],
    number
  >,
  "queryKey" | "getNextPageParam" | "initialPageParam"
>;
