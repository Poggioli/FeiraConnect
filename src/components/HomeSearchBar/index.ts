export * from "./HomeSearchBar";
export * from "./useHomeSearchBar";
export * from "./types";