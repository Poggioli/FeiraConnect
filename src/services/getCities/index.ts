export * from "./getCities";
export * from "./types";
