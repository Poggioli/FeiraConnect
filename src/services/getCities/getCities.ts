import { DEFAULT_PER_PAGE, api } from "@/services/api";
import { InfiniteData, useInfiniteQuery } from "@tanstack/react-query";
import { AxiosError, AxiosRequestConfig } from "axios";
import qs from "qs";
import { useMemo } from "react";
import {
  GetCities,
  GetCitiesParams,
  GetCitiesResponse,
  UseGetCitiesOptions,
} from "./types";

const url = 'cities' as const

async function fetchGetCities(
  { page, perPage, q }: GetCitiesParams,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  config: AxiosRequestConfig<any> | undefined = {}
): Promise<GetCitiesResponse> {
  const params = qs.stringify({ page, perPage, q });
  const res = await api.get<GetCitiesResponse>(
    `${url}?${params}`,
    config
  );
  return res.data;
}

const queryKey = "getCities" as const;

export function createUseGetCitiesQueryKey(query: string): ["getCities", string] {
  return [queryKey, query];
}

export function useGetCities(
  query: string,
  options?: UseGetCitiesOptions
) {
  const result = useInfiniteQuery<
    GetCitiesResponse,
    AxiosError,
    InfiniteData<GetCitiesResponse>,
    ["getCities", string],
    number
  >({
    ...options,
    queryKey: createUseGetCitiesQueryKey(query),
    initialPageParam: 1,
    queryFn: ({ pageParam = 1, signal }) =>
      fetchGetCities(
        {
          page: pageParam,
          perPage: DEFAULT_PER_PAGE,
          q: query
        },
        {
          signal,
        }
      ),
    getNextPageParam: (lastPage) =>
      lastPage.metadata.isLastPage
        ? undefined
        : lastPage.metadata.currentPage + 1,
  });

  const dataItems: GetCities[] = useMemo(
    () =>
      (result.data?.pages || []).reduce(
        (acc: GetCities[], page: GetCitiesResponse) => [
          ...acc,
          ...page.items,
        ],
        []
      ),
    [result.data]
  );

  const isEmpty: boolean = useMemo(() => !dataItems.length, [dataItems]);
  const isFilled: boolean = useMemo(() => !!dataItems.length, [dataItems]);

  return { ...result, dataItems, isEmpty, isFilled };
}
