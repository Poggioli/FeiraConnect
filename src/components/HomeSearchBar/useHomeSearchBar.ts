import { isEnterKey } from "@/lib/keyboardEvents";
import { useGetCities } from "@/services/getCities";
import { useNavigate, useSearch } from "@tanstack/react-router";
import { ChangeEvent, KeyboardEvent, useEffect, useState } from "react";

export function useHomeSearchBar() {
  const { q = "" } = useSearch({
    from: "/",
  });
  const navigate = useNavigate({ from: "/" });
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const citiesService = useGetCities(q, { enabled: false });

  function handleSearch(): void {
    if (q) {
      setIsOpen(true);
      citiesService.fetchNextPage();
    }
  }

  function handleOnChangeSearchValue(
    event: ChangeEvent<HTMLInputElement>
  ): void {
    navigate({
      search: {
        q: event.target.value,
      },
      replace: true,
    });
  }

  function handleOnKeyPress(event: KeyboardEvent<HTMLInputElement>): void {
    if (isEnterKey(event)) {
      handleSearch();
    }
  }

  useEffect(() => {
    if (q) {
      handleSearch();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return {
    searchValue: q,
    handleSearch,
    handleOnChangeSearchValue,
    handleOnKeyPress,
    onOpenChange: setIsOpen,
    isOpen,
    citiesService,
  };
}
