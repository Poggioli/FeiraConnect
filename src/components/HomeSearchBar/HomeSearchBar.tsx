import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Popover, PopoverAnchor, PopoverContent } from "@/components/ui/popover";
import { ScrollArea } from "@/components/ui/scroll-area";
import { Skeleton } from "@/components/ui/skeleton";
import { Search } from "lucide-react";
import { FC, useCallback, useRef } from "react";
import { HomeSearchBarProps } from "./types";
import { useHomeSearchBar } from "./useHomeSearchBar";

export const HomeSearchBar: FC<HomeSearchBarProps> = () => {

  const observer = useRef<IntersectionObserver>();
  const {
    handleOnChangeSearchValue,
    handleOnKeyPress,
    handleSearch,
    onOpenChange,
    searchValue,
    isOpen,
    citiesService: {
      isFetching,
      isSuccess,
      isError,
      isEmpty,
      isFilled,
      isLoading,
      hasNextPage,
      dataItems,
      fetchNextPage,
      refetch
    }
  } = useHomeSearchBar();

  const lastElementRef = useCallback((node: HTMLParagraphElement) => {
    if (isLoading) {
      return;
    }

    if (observer.current) {
      observer.current.disconnect();
    }

    observer.current = new IntersectionObserver((entries) => {
      if (entries[0].isIntersecting && hasNextPage && !isFetching) {
        fetchNextPage();
      }
    });

    if (node) {
      observer.current.observe(node);
    }
  }, [fetchNextPage, hasNextPage, isFetching, isLoading]);

  function handleOnClickTryAgain() {
    refetch();
  }

  return (
    <div className="w-full max-w-screen-sm flex flex-row gap-1">
      <Popover open={isOpen} onOpenChange={onOpenChange}>
        <PopoverAnchor asChild>
          <Input
            placeholder="Busque a cidade"
            value={searchValue}
            onChange={handleOnChangeSearchValue}
            onKeyDown={handleOnKeyPress}
          />
        </PopoverAnchor>
        <PopoverContent className="flex flex-col items-start justify-center gap-2 w-[var(--radix-popper-anchor-width)]">
          {isLoading ? (
            Array.from(Array(3).keys()).map((item) =>
              <div key={item} className="flex flex-col justify-center h-11">
                <Skeleton key={item} className="h-6 w-28" />
              </div>
            )
          ) : null}
          {!isLoading && isSuccess && isFilled ? (
            <ScrollArea className="w-full h-[calc(var(--radix-popover-content-available-height)-3rem)] overflow-hidden" >
              <div className="flex flex-col p-1">
                {dataItems.map((city) => (
                  <p
                    key={city.id}
                    ref={lastElementRef}
                    className="p-2 leading-7 text-muted-foreground hover:bg-muted hover:text-primary rounded-sm"
                  >
                    {city.name}
                  </p>
                ))}
                {isFetching ? (
                  Array.from(Array(3).keys()).map((item) =>
                    <div key={item} className="flex flex-col justify-center h-11">
                      <Skeleton key={item} className="h-6 w-28" />
                    </div>
                  )
                ) : null}
              </div>
            </ScrollArea>
          ) : null}
          {!isLoading && isSuccess && isEmpty ? (
            <div className="text-center w-full flex flex-col gap-1 justify-center items-center">
              <p className="leading-6 text-muted-foreground">Não encontramos essa cidade que você busca</p>
            </div>
          ) : null}
          {!isFetching && isError ? (
            <div className="text-center w-full flex flex-col gap-1 justify-center items-center">
              <p className="leading-6 text-muted-foreground">Aconteceu algum erro indo para feira...</p>
              <Button size="default" variant="link" onClick={handleOnClickTryAgain}>Tentar novamente</Button>
            </div>
          ) : null}
        </PopoverContent>
      </Popover>
      <Button
        variant="outline"
        size="icon"
        aria-label="Buscar cidade"
        onClick={handleSearch}
      >
        <Search className="w-4 h-4" />
      </Button>
    </div>
  );
};
