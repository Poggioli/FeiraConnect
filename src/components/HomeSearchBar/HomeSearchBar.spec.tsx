import { waitFor } from "@testing-library/react";
import { customRender } from "@tests/customRender";
import { HomeSearchBar } from "./HomeSearchBar";
import { server } from "@tests/server.mock";
import { getCitiesHandlers } from "@tests/getCities.mock";

const mockUseSearch = vi.fn();

vi.mock("@tanstack/react-router", async (importOriginal) => {
  const original = await importOriginal<Record<string, unknown>>();
  return {
    ...original,
    useSearch: (args: unknown) => mockUseSearch(args),
  };
});

describe("@/components/HomeSearchBar", () => {

  beforeEach(() => {
    mockUseSearch.mockReturnValue({});
  });

  it(`GIVEN a HomeSearchBar
      WHEN render
      THEN should show default state`, async () => {
    const { getByPlaceholderText, getByLabelText } = customRender(<HomeSearchBar />)

    await waitFor(() => {
      const input = getByPlaceholderText("Busque a cidade");
      const searchButton = getByLabelText("Buscar cidade");

      expect(input).toBeVisible();
      expect(searchButton).toBeVisible();
    });
  });

  it(`GIVEN a HomeSearchBar with search param
      WHEN service return a filled response
      THEN should show the link items`, async () => {
    mockUseSearch.mockReturnValue({ q: "test" });
    server.use(getCitiesHandlers.filledMultPagesPageOne);
    const { getByRole, getAllByText } = customRender(<HomeSearchBar />)

    await waitFor(() => {
      const popover = getByRole("dialog");
      const links = getAllByText(/Slug\s(\d)+/);

      expect(popover).toBeVisible();
      expect(links).toHaveLength(10);
    });
  });

  it(`GIVEN a HomeSearchBar with search param
      WHEN service return a empty response
      THEN should show the empty message`, async () => {
    mockUseSearch.mockReturnValue({ q: "test" });
    server.use(getCitiesHandlers.empty);
    const { getByText } = customRender(<HomeSearchBar />)

    await waitFor(() => {
      const emptyMessage = getByText("NÃ£o encontramos nenhuma cidade com este nome, entre em contato conosco.");

      expect(emptyMessage).toBeVisible();
    });
  });

  it(`GIVEN a HomeSearchBar with search param
      WHEN service return an error
      THEN should show the error message`, async () => {
    mockUseSearch.mockReturnValue({ q: "test" });
    server.use(getCitiesHandlers.error);
    const { getByText } = customRender(<HomeSearchBar />)

    await waitFor(() => {
      const errorMessage = getByText("Ocorreu um erro ao obter os dados. Tente novamente mais tarde.");
      const errorButton = getByText("Tentar novamente");

      expect(errorMessage).toBeVisible();
      expect(errorButton).toBeVisible();
    });
  });

  it(`GIVEN a HomeSearchBar with search param
      WHEN service is in loading state
      THEN should show the Skleton elements
      AND have aria-busy as true`, async () => {
    mockUseSearch.mockReturnValue({ q: "test" });
    server.use(getCitiesHandlers.filledOnePageWithDelay);
    const { getByRole, baseElement } = customRender(<HomeSearchBar />);

    await waitFor(() => {
      const popover = getByRole("dialog");
      expect(popover).toHaveAttribute("aria-busy", "true");
      expect(baseElement).toMatchSnapshot();
    })
  })
});