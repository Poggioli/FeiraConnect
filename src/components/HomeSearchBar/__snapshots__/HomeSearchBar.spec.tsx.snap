// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@/components/HomeSearchBar > GIVEN a HomeSearchBar with search param
      WHEN service is in loading state
      THEN should show the Skleton elements
      AND have aria-busy as true 1`] = `
<body>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <div
      class="w-full max-w-screen-sm flex flex-row gap-1"
    >
      <input
        class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
        placeholder="Busque a cidade"
        value="test"
      />
      <button
        aria-label="Buscar cidade"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 w-9"
      >
        <svg
          class="lucide lucide-search w-4 h-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <path
            d="m21 21-4.3-4.3"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    style="position: fixed; left: 0px; top: 0px; transform: translate(0, -200%); min-width: max-content;"
  >
    <div
      aria-busy="true"
      class="z-50 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 flex flex-col items-start justify-center gap-2 w-[var(--radix-popper-anchor-width)]"
      data-align="center"
      data-side="bottom"
      data-state="open"
      id="radix-:r4:"
      role="dialog"
      style="--radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height); animation: none;"
      tabindex="-1"
    >
      <div
        class="flex flex-col justify-center h-11"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-6 w-28"
        />
      </div>
      <div
        class="flex flex-col justify-center h-11"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-6 w-28"
        />
      </div>
      <div
        class="flex flex-col justify-center h-11"
      >
        <div
          class="animate-pulse rounded-md bg-primary/10 h-6 w-28"
        />
      </div>
    </div>
  </div>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
